shell/platform/embedder/tests/embedder_test.h

```diff
@@ -16,29 +16,63 @@
 
 namespace flutter::testing {
 
+class EmbedderTestContextGL;
+class EmbedderTestContextMetal;
+class EmbedderTestContextSoftware;
+class EmbedderTestContextVulkan;
+
 class EmbedderTest : public ThreadTest {
  public:
   EmbedderTest();
 
   std::string GetFixturesDirectory() const;
 
-  EmbedderTestContext& GetEmbedderContext(EmbedderTestContextType type);
+  template <typename T>
+  T& GetEmbedderContext() {
+    static_assert(false, "Unsupported test context type");
+  }
+
+  template <>
+  EmbedderTestContextGL& GetEmbedderContext<EmbedderTestContextGL>() {
+    return reinterpret_cast<EmbedderTestContextGL&>(GetGLContext());
+  }
+
+  template <>
+  EmbedderTestContextMetal& GetEmbedderContext<EmbedderTestContextMetal>() {
+    return reinterpret_cast<EmbedderTestContextMetal&>(GetMetalContext());
+  }
+
+  template <>
+  EmbedderTestContextSoftware&
+  GetEmbedderContext<EmbedderTestContextSoftware>() {
+    return reinterpret_cast<EmbedderTestContextSoftware&>(GetSoftwareContext());
+  }
 
- private:
-  std::map<EmbedderTestContextType, std::unique_ptr<EmbedderTestContext>>
-      embedder_contexts_;
+  template <>
+  EmbedderTestContextVulkan& GetEmbedderContext<EmbedderTestContextVulkan>() {
+    return reinterpret_cast<EmbedderTestContextVulkan&>(GetVulkanContext());
+  }
 
-  std::unique_ptr<EmbedderTestContext> CreateSoftwareContext();
-  std::unique_ptr<EmbedderTestContext> CreateGLContext();
-  std::unique_ptr<EmbedderTestContext> CreateMetalContext();
-  std::unique_ptr<EmbedderTestContext> CreateVulkanContext();
+ protected:
+  EmbedderTestContext& GetGLContext();
+  EmbedderTestContext& GetMetalContext();
+  EmbedderTestContext& GetSoftwareContext();
+  EmbedderTestContext& GetVulkanContext();
+
+  std::unique_ptr<EmbedderTestContext> gl_context_;
+  std::unique_ptr<EmbedderTestContext> metal_context_;
+  std::unique_ptr<EmbedderTestContext> software_context_;
+  std::unique_ptr<EmbedderTestContext> vulkan_context_;
```

Comment:
Unfortunately these need to be base class types since the subclasses are only forward-declared here, not `#include`d. We don't include them because it'd be nice to use ARC-managed Obj-C types in the Metal headers, but pulling those into pure C++ translation units makes them sad.